pkg/controller/infrastructure/actuator_delete.go

```diff
@@ -27,15 +27,35 @@ import (
 	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
 	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
 	"github.com/gardener/gardener/pkg/utils/flow"
+	versionutils "github.com/gardener/gardener/pkg/utils/version"
 	"github.com/go-logr/logr"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/client-go/rest"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func (a *actuator) Delete(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, _ *extensionscontroller.Cluster) error {
+func (a *actuator) Delete(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster) error {
 	logger := a.logger.WithValues("infrastructure", client.ObjectKeyFromObject(infrastructure), "operation", "delete")
-	return Delete(ctx, logger, a.RESTConfig(), a.Client(), infrastructure)
+
+	// Prior to Kubernetes v1.16 there was no finalizer support for `Service` objects. This is especially painful for
+	// `Service`s of type `LoadBalancer` because infrastructure artefacts are created. An deletion of the `Service`
+	// object leads to immediate disappearance from the system/cluster while the service-controller is still processing
+	// the deletion of the infrastructure artefacts in the background. Consequently, as an end-user you cannot know when
+	// the service-controller finished its work and cleaned up properly.
+	// If it didn't finish the destruction fast enough then the infrastructure deletion (trying to destroy the subnets,
+	// VPC, etc.) will fail due to the fact that there are still resources (load balancers, security groups) that we
+	// are not aware of. As a result, manual intervention is required to cleanup those artefacts to allow our infrastructure
+	// deletion to properly proceed.
+	// To mitigate this, we are explicitly deletion load balancers and security groups for Kubernetes versions lower
+	// than v1.16 (in fact, we are duplicating the logic of the service-controller here (at least parts of it)).
+	// See more details: https://github.com/kubernetes/enhancements/issues/980
+	// TODO: Remove all this code when the extension does only support clusters with at least Kubernetes v1.16.
+	needsExplicitLoadBalancerDeletion, err := versionutils.CheckVersionMeetsConstraint(cluster.Shoot.Spec.Kubernetes.Version, "< 1.16")
```

Comment:
In the integration tests the created Cluster resource is without any Shoot set, I guess that is why the test failed with nil pointer dereference in this line.

Reply:
Yep, I'm on it.

Reply:
I have adapted the `Cluster` resource now (inspired from https://github.com/gardener/gardener-extension-provider-aws/blob/6fb9da86dbd2f88413403175599bb871bc64db8a/pkg/controller/controlplane/valuesprovider_test.go#L68-L92). WDYT?